src/optimal_retention.rs

```diff
@@ -206,251 +161,163 @@ pub fn simulate(
         }
 
-        // dbg!(&cum_sum);
+        let ivl = next_interval(card.stability, desired_retention)
+            .round()
+            .clamp(1.0, max_ivl);
 
-        // Create 'true_learn' mask based on 'need_learn' and 'cum_sum' and 'learn_limit'
-        let mut learn_count = 0;
-        let true_learn =
-            Zip::from(&need_learn)
-                .and(&cum_sum)
-                .map_collect(|&need_learn_flag, &cum_cost| {
-                    if need_learn_flag {
-                        learn_count += 1;
-                    }
-                    need_learn_flag && (cum_cost <= max_cost_perday) && (learn_count <= learn_limit)
-                });
+        // dbg!(ivl);
 
-        let mut new_stability = old_stability.to_owned();
-        let old_difficulty = card_table.slice(s![Column::Difficulty, ..]);
-        // Iterate over slices and apply stability_after_failure function
-        izip!(
-            &mut new_stability,
-            &old_stability,
-            &retrievability,
-            &old_difficulty,
-            &(&true_review & &forget)
-        )
-        .filter(|(.., &condition)| condition)
-        .for_each(|(new_stab, &stab, &retr, &diff, ..)| {
-            let post_lapse_stab = stability_after_failure(w, stab, retr, diff);
-            *new_stab =
-                stability_short_term(w, post_lapse_stab, forget_rating_offset, forget_session_len);
-        });
+        // Update days statistics
 
-        // Iterate over slices and apply stability_after_success function
-        izip!(
-            &mut new_stability,
-            &ratings,
-            &old_stability,
-            &retrievability,
-            &old_difficulty,
-            &(&true_review & !&forget)
-        )
-        .filter(|(.., &condition)| condition)
-        .for_each(|(new_stab, &rating, &stab, &retr, &diff, _)| {
-            *new_stab = stability_after_success(w, stab, retr, diff, rating);
-        });
+        if not_learn {
+            review_cnt_per_day[day_index] += 1;
+        } else {
+            learn_cnt_per_day[day_index] += 1
+        }
+        cost_per_day[day_index] += cost;
 
-        // Initialize a new Array1 to store updated difficulty values
-        let mut new_difficulty = old_difficulty.to_owned();
+        // dbg!(&review_cnt_per_day);
 
-        // Update difficulty for review cards
-        izip!(&mut new_difficulty, &old_difficulty, &ratings, &true_review,)
-            .filter(|(.., &condition)| condition)
-            .for_each(|(new_diff, &old_diff, &rating, ..)| {
-                *new_diff = next_d(w, old_diff, rating);
-                if rating == 1 {
-                    *new_diff -= (w[6] * forget_rating_offset).clamp(1.0, 10.0);
-                }
-            });
+        let upper = min(day_index + ivl as usize, learn_span);
+        for i in day_index..upper {
+            memorized_cnt_per_day[i] += retrievability;
+        }
```

Comment:
I figure it out. The `retrievability` should decay over time.

Reply:
left: 6475.4844
 right: 6521.068

Oh yeah that's a lot better.